SLS-1691: Add Source Code Integration

IvanTopolcic opened the pull request:
<!--- Please remember to review the [contribution guidelines](https://github.com/DataDog/serverless-plugin-datadog/blob/master/CONTRIBUTING.md) if you have not yet done so._  --->

### What does this PR do?

You only need to review what's in the root directory, most of the code is pulled from datadog-ci.

This PR adds an option to the serverless plugin to enable source code integration. By default, this flag is enabled.

### Motivation

This PR is part of source code integration.

### Testing Guidelines

This PR was tested by hand and has unit tests associated with it. I tested the flow for instrumenting source code integration on a few different lambda functions.

### Additional Notes

Note that for our customers this only works in Go and Java for now.

### Types of changes

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [x] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [x] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
-| `customHandler`             | When set, the specified handler is set as the handler for all the functions. By deafult, the handler is set to `/opt/nodejs/node_modules/datadog-lambda-js/handler.handler` if `addLayers` is set to `true`, or `node_modules/datadog-lambda-js/dist/handler.handler` if `addLayers` is set to `false`                                                                          ...
+| Parameter                     | Description                                                                                                                                                                                                                                                                                                                                                                   ...
+| ----------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| `flushMetricsToLogs`          | Send custom metrics by using logs with the Datadog Forwarder Lambda function (recommended). Defaults to `true`. If you disable this parameter, it's required to set `DATADOG_API_KEY` in your environment, or if encrypted, `apiKMSKey` or `apiKeySecretArn` in the configuration. `flushMetricsToLogs` is ignored when `addExtension` is true.                               ...
```
```suggestion
| `flushMetricsToLogs`          | Send custom metrics by using logs with the Datadog Forwarder Lambda function (recommended). Defaults to `true`. If you disable this parameter, you must set `DATADOG_API_KEY` in your environment. To use an encrypted key, set `apiKMSKey` or `apiKeySecretArn` in the configuration. `flushMetricsToLogs` is ignored when `addExtension` is true.                                                                                          |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| Parameter                     | Description                                                                                                                                                                                                                                                                                                                                                                   ...
+| ----------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| `flushMetricsToLogs`          | Send custom metrics by using logs with the Datadog Forwarder Lambda function (recommended). Defaults to `true`. If you disable this parameter, it's required to set `DATADOG_API_KEY` in your environment, or if encrypted, `apiKMSKey` or `apiKeySecretArn` in the configuration. `flushMetricsToLogs` is ignored when `addExtension` is true.                               ...
+| `site`                        | Set which Datadog site to send data, this is only used when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Possible values are `datadoghq.com`, `datadoghq.eu`, `us3.datadoghq.com`, `us5.datadoghq.com`, and `ddog-gov.com`. The default is `datadoghq.com`.                                                                                                   ...
```
```suggestion
| `site`                        | Set which Datadog site to send data to. This is only used when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Possible values are `datadoghq.com`, `datadoghq.eu`, `us3.datadoghq.com`, `us5.datadoghq.com`, and `ddog-gov.com`. The default is `datadoghq.com`.                                                                                                                                                              |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| ----------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| `flushMetricsToLogs`          | Send custom metrics by using logs with the Datadog Forwarder Lambda function (recommended). Defaults to `true`. If you disable this parameter, it's required to set `DATADOG_API_KEY` in your environment, or if encrypted, `apiKMSKey` or `apiKeySecretArn` in the configuration. `flushMetricsToLogs` is ignored when `addExtension` is true.                               ...
+| `site`                        | Set which Datadog site to send data, this is only used when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Possible values are `datadoghq.com`, `datadoghq.eu`, `us3.datadoghq.com`, `us5.datadoghq.com`, and `ddog-gov.com`. The default is `datadoghq.com`.                                                                                                   ...
+| `apiKey`                      | Datadog API Key, only needed when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Not required when DATADOG_API_KEY environment variable is set. Defining `apiKey` will add the Datadog API key directly to your Lambda functions as an environment variable. For more information about getting a Datadog API key, see the [API key documentation][3].          ...
```
```suggestion
| `apiKey`                      | Datadog API key. Only needed when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Not required when the DATADOG_API_KEY environment variable is set. Defining `apiKey` adds the Datadog API key directly to your Lambda functions as an environment variable. For more information about getting a Datadog API key, see the [API key documentation][3].                                                                     |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `flushMetricsToLogs`          | Send custom metrics by using logs with the Datadog Forwarder Lambda function (recommended). Defaults to `true`. If you disable this parameter, it's required to set `DATADOG_API_KEY` in your environment, or if encrypted, `apiKMSKey` or `apiKeySecretArn` in the configuration. `flushMetricsToLogs` is ignored when `addExtension` is true.                               ...
+| `site`                        | Set which Datadog site to send data, this is only used when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Possible values are `datadoghq.com`, `datadoghq.eu`, `us3.datadoghq.com`, `us5.datadoghq.com`, and `ddog-gov.com`. The default is `datadoghq.com`.                                                                                                   ...
+| `apiKey`                      | Datadog API Key, only needed when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Not required when DATADOG_API_KEY environment variable is set. Defining `apiKey` will add the Datadog API key directly to your Lambda functions as an environment variable. For more information about getting a Datadog API key, see the [API key documentation][3].          ...
+| `appKey`                      | Datadog App Key, only needed when `monitors` is defined. Not required when DATADOG_APP_KEY environment variable is set.                                                                                                                                                                                                                                                       ...
```
```suggestion
| `appKey`                      | Datadog app key. Only needed when `monitors` is defined. Not required when the DATADOG_APP_KEY environment variable is set.                                                                                                                                                                                                                                                                                                                  |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `apiKey`                      | Datadog API Key, only needed when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Not required when DATADOG_API_KEY environment variable is set. Defining `apiKey` will add the Datadog API key directly to your Lambda functions as an environment variable. For more information about getting a Datadog API key, see the [API key documentation][3].          ...
+| `appKey`                      | Datadog App Key, only needed when `monitors` is defined. Not required when DATADOG_APP_KEY environment variable is set.                                                                                                                                                                                                                                                       ...
+| `apiKeySecretArn`             | The ARN of the secret storing the Datadog API key in AWS Secrets Manager. Use this parameter if you're storing your secrets in AWS Secrets Manager, and when `flushMetricsToLogs` is `false` or `addExtension` is `true`. If set, remember to add the `secretsmanager:GetSecretValue` permission to the Lambda execution role.                                                ...
+| `apiKMSKey`                   | Datadog API Key encrypted using KMS. Use this parameter if you're storing your secrets with KMS, and if `flushMetricsToLogs` is `false` or `addExtension` is `true`. Defining `apiKMSKey` will add the Datadog API Key directly to your Lambda functions as an environment variable.                                                                                          ...
```
```suggestion
| `apiKMSKey`                   | Datadog API key encrypted using KMS. Use this parameter if you're storing your secrets with KMS, and if `flushMetricsToLogs` is `false` or `addExtension` is `true`. Defining `apiKMSKey` adds the Datadog API Key directly to your Lambda functions as an environment variable.                                                                                                                                                     |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `appKey`                      | Datadog App Key, only needed when `monitors` is defined. Not required when DATADOG_APP_KEY environment variable is set.                                                                                                                                                                                                                                                       ...
+| `apiKeySecretArn`             | The ARN of the secret storing the Datadog API key in AWS Secrets Manager. Use this parameter if you're storing your secrets in AWS Secrets Manager, and when `flushMetricsToLogs` is `false` or `addExtension` is `true`. If set, remember to add the `secretsmanager:GetSecretValue` permission to the Lambda execution role.                                                ...
+| `apiKMSKey`                   | Datadog API Key encrypted using KMS. Use this parameter if you're storing your secrets with KMS, and if `flushMetricsToLogs` is `false` or `addExtension` is `true`. Defining `apiKMSKey` will add the Datadog API Key directly to your Lambda functions as an environment variable.                                                                                          ...
+| `captureLambdaPayload`        | (Experimental) This optional setting configures Datadog ingestion for incoming and outgoing AWS Lambda payloads. Function request and response values are added as part of outgoing APM spans sent to Datadog, and require Datadog APM to be configured on your function. Obfuscation of fields in the Lambda payload is available via the `replace_tags` block within `apm_c ...
```
```suggestion
| `captureLambdaPayload`        | (Experimental) This optional setting configures Datadog ingestion for incoming and outgoing AWS Lambda payloads. Function request and response values are added as part of outgoing APM spans sent to Datadog, and require Datadog APM to be configured on your function. You can obfuscate fields in the Lambda payload by using the `replace_tags` block within `apm_config` settings in [datadog.yaml.][12] Defaults to `false`. |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `apiKMSKey`                   | Datadog API Key encrypted using KMS. Use this parameter if you're storing your secrets with KMS, and if `flushMetricsToLogs` is `false` or `addExtension` is `true`. Defining `apiKMSKey` will add the Datadog API Key directly to your Lambda functions as an environment variable.                                                                                          ...
+| `captureLambdaPayload`        | (Experimental) This optional setting configures Datadog ingestion for incoming and outgoing AWS Lambda payloads. Function request and response values are added as part of outgoing APM spans sent to Datadog, and require Datadog APM to be configured on your function. Obfuscation of fields in the Lambda payload is available via the `replace_tags` block within `apm_c ...
+| `addLayers`                   | Whether to install the Datadog Lambda library as a layer. Defaults to `true`. Set to `false` when you plan to package the Datadog Lambda library to your function's deployment package on your own so that you can install a specific version of the Datadog Lambda library ([Python][4] or [Node.js][5]).                                                                    ...
+| `addExtension`                | Whether to install the Datadog Lambda Extension as a layer. Defaults to `false`. When enabled, it's required to set the environment variable `DATADOG_API_KEY`, or the configuration values `apiKMSKey` or `apiKeySecretArn`. Learn more about the [Datadog Lambda Extension Layer][8] in the documentation. **Note**: AWS only supports Lambda Extensions for [certain runti ...
```
```suggestion
| `addExtension`                | Whether to install the Datadog Lambda Extension as a layer. Defaults to `false`. When enabled, you must set either the environment variable `DATADOG_API_KEY`, or the configuration values `apiKMSKey` or `apiKeySecretArn`. Learn more about the [Datadog Lambda Extension Layer][8] in the documentation. **Note**: AWS only supports Lambda Extensions for [certain runtimes][9].                                                    |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `logLevel`                    | The log level, set to `DEBUG` for extended logging.                                                                                                                                                                                                                                                                                                                           ...
+| `enableXrayTracing`           | Set `true` to enable X-Ray tracing on the Lambda functions and API Gateway integrations. Defaults to `false`.                                                                                                                                                                                                                                                                 ...
+| `enableDDTracing`             | Enable Datadog tracing on the Lambda function. Note: This applies only to integrations using the Datadog Extension. Defaults to `true`.                                                                                                                                                                                                                                       ...
+| `enableDDLogs`                | Enable Datadog log collection for the Lambda function. Note: This setting has no effect on logs sent via the Datadog Forwarder. Defaults to `true`.                                                                                                                                                                                                                           ...
```
```suggestion
| `enableDDLogs`                | Enable Datadog log collection for the Lambda function. Note: This setting has no effect on logs sent by the Datadog Forwarder. Defaults to `true`.                                                                                                                                                                                                                                                                                      |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `enableDDLogs`                | Enable Datadog log collection for the Lambda function. Note: This setting has no effect on logs sent via the Datadog Forwarder. Defaults to `true`.                                                                                                                                                                                                                           ...
+| `enableSourceCodeIntegration` | Enable Datadog Source Code integration for the function. Only works for Java and Go.                                                                                                                                                                                                                                                                                          ...
+| `subscribeToApiGatewayLogs`   | Enable automatic subscription of the Datadog Forwarder to API Gateway log groups. Defaults to `true`.                                                                                                                                                                                                                                                                         ...
+| `subscribeToHttpApiLogs`      | Enable automatic subscription of the Datadog Forwarder to Http-api log groups. Defaults to `true`.                                                                                                                                                                                                                                                                            ...
```
```suggestion
| `subscribeToHttpApiLogs`      | Enable automatic subscription of the Datadog Forwarder to HTTP API log groups. Defaults to `true`.                                                                                                                                                                                                                                                                                                                                       |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `enableSourceCodeIntegration` | Enable Datadog Source Code integration for the function. Only works for Java and Go.                                                                                                                                                                                                                                                                                          ...
+| `subscribeToApiGatewayLogs`   | Enable automatic subscription of the Datadog Forwarder to API Gateway log groups. Defaults to `true`.                                                                                                                                                                                                                                                                         ...
+| `subscribeToHttpApiLogs`      | Enable automatic subscription of the Datadog Forwarder to Http-api log groups. Defaults to `true`.                                                                                                                                                                                                                                                                            ...
+| `subscribeToWebsocketLogs`    | Enable automatic subscription of the Datadog Forwarder to Websocket log groups. Defaults to `true`.                                                                                                                                                                                                                                                                           ...
```
```suggestion
| `subscribeToWebsocketLogs`    | Enable automatic subscription of the Datadog Forwarder to WebSocket log groups. Defaults to `true`.                                                                                                                                                                                                                                                                                                                                      |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `subscribeToApiGatewayLogs`   | Enable automatic subscription of the Datadog Forwarder to API Gateway log groups. Defaults to `true`.                                                                                                                                                                                                                                                                         ...
+| `subscribeToHttpApiLogs`      | Enable automatic subscription of the Datadog Forwarder to Http-api log groups. Defaults to `true`.                                                                                                                                                                                                                                                                            ...
+| `subscribeToWebsocketLogs`    | Enable automatic subscription of the Datadog Forwarder to Websocket log groups. Defaults to `true`.                                                                                                                                                                                                                                                                           ...
+| `forwarderArn`                | Setting this parameter subscribes the given Datadog forwarder to the Lambda functions’ CloudWatch log groups. Required when `enableDDTracing` is set to `true` unless the subscription is otherwise applied. For example, if a Datadog Forwarder subscription is applied via Datadog's AWS Integration, then `forwarderArn` is not required.                                  ...
```
```suggestion
| `forwarderArn`                | Setting this parameter subscribes the given Datadog forwarder to the Lambda functions’ CloudWatch log groups. Required when `enableDDTracing` is set to `true` unless the subscription is otherwise applied. For example, if a Datadog Forwarder subscription is applied using Datadog's AWS Integration, then `forwarderArn` is not required.                                                                                             |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `subscribeToHttpApiLogs`      | Enable automatic subscription of the Datadog Forwarder to Http-api log groups. Defaults to `true`.                                                                                                                                                                                                                                                                            ...
+| `subscribeToWebsocketLogs`    | Enable automatic subscription of the Datadog Forwarder to Websocket log groups. Defaults to `true`.                                                                                                                                                                                                                                                                           ...
+| `forwarderArn`                | Setting this parameter subscribes the given Datadog forwarder to the Lambda functions’ CloudWatch log groups. Required when `enableDDTracing` is set to `true` unless the subscription is otherwise applied. For example, if a Datadog Forwarder subscription is applied via Datadog's AWS Integration, then `forwarderArn` is not required.                                  ...
+| `integrationTesting`          | Set `true` when running integration tests. This will bypass the validation of the Forwarder ARN and the addition of Datadog Monitor output links. Defaults to `false`.                                                                                                                                                                                                        ...
```
```suggestion
| `integrationTesting`          | Set `true` when running integration tests. This bypasses the validation of the Forwarder ARN and the addition of Datadog Monitor output links. Defaults to `false`.                                                                                                                                                                                                                                                                   |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `forwarderArn`                | Setting this parameter subscribes the given Datadog forwarder to the Lambda functions’ CloudWatch log groups. Required when `enableDDTracing` is set to `true` unless the subscription is otherwise applied. For example, if a Datadog Forwarder subscription is applied via Datadog's AWS Integration, then `forwarderArn` is not required.                                  ...
+| `integrationTesting`          | Set `true` when running integration tests. This will bypass the validation of the Forwarder ARN and the addition of Datadog Monitor output links. Defaults to `false`.                                                                                                                                                                                                        ...
+| `enableTags`                  | When set, automatically tag the Lambda functions with the `service` and `env` tags using the `service` and `stage` values from the serverless application definition. It does NOT override if a `service` or `env` tag already exists. Defaults to `true`.                                                                                                                    ...
+| `injectLogContext`            | When set, the lambda layer will automatically patch console.log with Datadog's tracing ids. Defaults to `true`.                                                                                                                                                                                                                                                               ...
```
```suggestion
| `injectLogContext`            | When set, the Lambda layer automatically patches `console.log` with Datadog's tracing IDs. Defaults to `true`.                                                                                                                                                                                                                                                                                                                          |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `integrationTesting`          | Set `true` when running integration tests. This will bypass the validation of the Forwarder ARN and the addition of Datadog Monitor output links. Defaults to `false`.                                                                                                                                                                                                        ...
+| `enableTags`                  | When set, automatically tag the Lambda functions with the `service` and `env` tags using the `service` and `stage` values from the serverless application definition. It does NOT override if a `service` or `env` tag already exists. Defaults to `true`.                                                                                                                    ...
+| `injectLogContext`            | When set, the lambda layer will automatically patch console.log with Datadog's tracing ids. Defaults to `true`.                                                                                                                                                                                                                                                               ...
+| `exclude`                     | When set, this plugin will ignore all specified functions. Use this parameter if you have any functions that should not include Datadog functionality. Defaults to `[]`.                                                                                                                                                                                                      ...
```
```suggestion
| `exclude`                     | When set, this plugin ignores all specified functions. Use this parameter if you have any functions that should not include Datadog functionality. Defaults to `[]`.                                                                                                                                                                                                                                                                 |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `enableTags`                  | When set, automatically tag the Lambda functions with the `service` and `env` tags using the `service` and `stage` values from the serverless application definition. It does NOT override if a `service` or `env` tag already exists. Defaults to `true`.                                                                                                                    ...
+| `injectLogContext`            | When set, the lambda layer will automatically patch console.log with Datadog's tracing ids. Defaults to `true`.                                                                                                                                                                                                                                                               ...
+| `exclude`                     | When set, this plugin will ignore all specified functions. Use this parameter if you have any functions that should not include Datadog functionality. Defaults to `[]`.                                                                                                                                                                                                      ...
+| `enabled`                     | When set to false, the Datadog plugin will stay inactive. Defaults to `true`. You can control this option using an environment variable, e.g. `enabled: ${strToBool(${env:DD_PLUGIN_ENABLED, true})}`, to activate/deactivate the plugin during deployment. Alternatively, you can also use the value passed in through `--stage` to control this option, [see example.](#dis ...
```
```suggestion
| `enabled`                     | When set to `false`, the Datadog plugin stays inactive. Defaults to `true`. You can control this option using an environment variable. For example, use `enabled: ${strToBool(${env:DD_PLUGIN_ENABLED, true})}` to activate/deactivate the plugin during deployment. Alternatively, you can also use the value passed in through `--stage` to control this option—[see example](#disable-plugin-for-particular-environment).                     |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `injectLogContext`            | When set, the lambda layer will automatically patch console.log with Datadog's tracing ids. Defaults to `true`.                                                                                                                                                                                                                                                               ...
+| `exclude`                     | When set, this plugin will ignore all specified functions. Use this parameter if you have any functions that should not include Datadog functionality. Defaults to `[]`.                                                                                                                                                                                                      ...
+| `enabled`                     | When set to false, the Datadog plugin will stay inactive. Defaults to `true`. You can control this option using an environment variable, e.g. `enabled: ${strToBool(${env:DD_PLUGIN_ENABLED, true})}`, to activate/deactivate the plugin during deployment. Alternatively, you can also use the value passed in through `--stage` to control this option, [see example.](#dis ...
+| `monitors`                    | When defined, the Datadog plugin will configure monitors for the deployed function. Requires setting `DATADOG_API_KEY` and `DATADOG_APP_KEY` in your environment. To learn how to define monitors, see [To Enable and Configure a Recommended Serverless Monitor.](#to-enable-and-configure-a-recommended-serverless-monitor)                                                 ...
```
```suggestion
| `monitors`                    | When defined, the Datadog plugin will configure monitors for the deployed function. Requires setting `DATADOG_API_KEY` and `DATADOG_APP_KEY` in your environment. To learn how to define monitors, see [To Enable and Configure a Recommended Serverless Monitor](#to-enable-and-configure-a-recommended-serverless-monitor).                                                                                                            |
```

cswatt commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `exclude`                     | When set, this plugin will ignore all specified functions. Use this parameter if you have any functions that should not include Datadog functionality. Defaults to `[]`.                                                                                                                                                                                                      ...
+| `enabled`                     | When set to false, the Datadog plugin will stay inactive. Defaults to `true`. You can control this option using an environment variable, e.g. `enabled: ${strToBool(${env:DD_PLUGIN_ENABLED, true})}`, to activate/deactivate the plugin during deployment. Alternatively, you can also use the value passed in through `--stage` to control this option, [see example.](#dis ...
+| `monitors`                    | When defined, the Datadog plugin will configure monitors for the deployed function. Requires setting `DATADOG_API_KEY` and `DATADOG_APP_KEY` in your environment. To learn how to define monitors, see [To Enable and Configure a Recommended Serverless Monitor.](#to-enable-and-configure-a-recommended-serverless-monitor)                                                 ...
+| `customHandler`               | When set, the specified handler is set as the handler for all the functions. By deafult, the handler is set to `/opt/nodejs/node_modules/datadog-lambda-js/handler.handler` if `addLayers` is set to `true`, or `node_modules/datadog-lambda-js/dist/handler.handler` if `addLayers` is set to `false`                                                                        ...
```
```suggestion
| `customHandler`               | When set, the specified handler is set as the handler for all the functions. By default, the handler is set to `/opt/nodejs/node_modules/datadog-lambda-js/handler.handler` if `addLayers` is set to `true`, or `node_modules/datadog-lambda-js/dist/handler.handler` if `addLayers` is set to `false`.                                                                                                                                   |
```

cswatt reviewed (changes requested):
just readme grammar changes

nhinsch commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `enableXrayTracing`           | Set `true` to enable X-Ray tracing on the Lambda functions and API Gateway integrations. Defaults to `false`.                                                                                                                                                                                                                                                                 ...
+| `enableDDTracing`             | Enable Datadog tracing on the Lambda function. Note: This applies only to integrations using the Datadog Extension. Defaults to `true`.                                                                                                                                                                                                                                       ...
+| `enableDDLogs`                | Enable Datadog log collection for the Lambda function. Note: This setting has no effect on logs sent by the Datadog Forwarder. Defaults to `true`.                                                                                                                                                                                                                            ...
+| `enableSourceCodeIntegration` | Enable Datadog Source Code integration for the function. Only works for Java and Go.                                                                                                                                                                                                                                                                                          ...
```
1. Source Code -> source code
2. Can you elaborate on what this actually means/does?
3. For "only works for Java and Go", I would rephrase this slightly to be a bit softer, maybe "This feature is currently only supported for functions written in Java or Go."

IvanTopolcic replied:
Good idea, I actually originally had this but removed it because I felt it was a bit much and couldn't figure out a way to word it. Will do!

nhinsch commented on src/git-metadata/api.ts:
```diff
@@ -0,0 +1,11 @@
+export const datadogSite = process.env.DATADOG_SITE || "datadoghq.com";
```
If I understand correctly, this means that the DATADOG_SITE must be set using an environment, variable and we don't use the `site` parameter from the config file. What is the rationale for that?

If we do have a separate environment variable/setting for this, we probably need to document that, and also validate that it's one of the valid values (e.g. not misspelled).

IvanTopolcic replied:
Good catch, this was just leftover cruft from datadog-ci -- I'll use the `site` that we use in serverless.

nhinsch commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
-| `customHandler`             | When set, the specified handler is set as the handler for all the functions. By deafult, the handler is set to `/opt/nodejs/node_modules/datadog-lambda-js/handler.handler` if `addLayers` is set to `true`, or `node_modules/datadog-lambda-js/dist/handler.handler` if `addLayers` is set to `false`                                                                          ...
+| Parameter                     | Description                                                                                                                                                                                                                                                                                                                                                                   ...
+| ----------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| `flushMetricsToLogs`          | Send custom metrics by using logs with the Datadog Forwarder Lambda function (recommended). Defaults to `true`. If you disable this parameter, you must set `DATADOG_API_KEY` in your environment. To use an encrypted key, set `apiKMSKey` or `apiKeySecretArn` in the configuration. `flushMetricsToLogs` is ignored when `addExtension` is true.                           ...
```
> If you disable this parameter, you must set `DATADOG_API_KEY` in your environment.

Could you also just set the `apiKey` config value?

IvanTopolcic replied:
That's right, just `apiKey` would work.

nhinsch commented on README.md:
```diff
@@ -31,32 +31,33 @@ Each version of the plugin is published with a [specific set of versions of the
+| `apiKey`                      | Datadog API key. Only needed when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Not required when the DATADOG_API_KEY environment variable is set. Defining `apiKey` adds the Datadog API key directly to your Lambda functions as an environment variable. For more information about getting a Datadog API key, see the [API key documentation][3].          ...
+| `appKey`                      | Datadog app key. Only needed when `monitors` is defined. Not required when the DATADOG_APP_KEY environment variable is set.                                                                                                                                                                                                                                                   ...
+| `apiKeySecretArn`             | The ARN of the secret storing the Datadog API key in AWS Secrets Manager. Use this parameter if you're storing your secrets in AWS Secrets Manager, and when `flushMetricsToLogs` is `false` or `addExtension` is `true`. If set, remember to add the `secretsmanager:GetSecretValue` permission to the Lambda execution role.                                                ...
+| `apiKMSKey`                   | Datadog API key encrypted using KMS. Use this parameter if you're storing your secrets with KMS, and if `flushMetricsToLogs` is `false` or `addExtension` is `true`. Defining `apiKMSKey` adds the Datadog API Key directly to your Lambda functions as an environment variable.                                                                                              ...
```
I think the different types of API key settings have gotten a bit confusing. It could be a good idea to clarify the situation a bit, maybe reorganize this part to try to make things more clear.

nhinsch commented:
You mentioned that the feature only works for Java and Go, but I don't see any logic related to checking for those runtimes, is that to be expected?

nhinsch commented:
I don't think this should block the PR, but if we want to implement this feature in more places I think it would make sense to extract the code into a reusable library that we can share across our deployment tools (including the CI).

IvanTopolcic commented:
Thanks for the review Nick. I've gone ahead and fixed the datadogSite issue and attempted to make the documentation a bit clearer. I didn't change it too much so let me know if you think it's still a bit hard to understand, and if so, what you think we should change (imo I think repeating the `addExtension` is true and `flushMetricsToLogs` is false thing is kind of repetitive and confusing).

As for the other two comments:

(1) It only works for Java and Go but it was recommended by the Source Code Integration team to leave it on for any repo as eventually it'll be supported for most languages. I also removed Java and Go from the description of the feature flag and instead linked to the documentation on Source Code Integration, which includes a message stating that the feature is in public beta and only works for Go and JVM based languages.
(2) This is actually actively being worked on by the Source Code Integration team and they have a jira ticket open for this. Whenever they get around to it we can easily import datadog-ci as a dependency and use it. Could be good to do for a future engineering sprint.

nhinsch commented on src/index.ts:
```diff
@@ -142,6 +159,30 @@ module.exports = class ServerlessPlugin {
+    // We only want to enable source code integration if the following conditions are met:
+    // 1. The flag is enabled (it's enabled by default)
+    // 2. They've provided a DATADOG_API_KEY (this will silently fail if they do not provide an API key)
+    // 3. They're in a Git repo (this will silently fail if they do not provide an API key)
```
I don't think you need this comment here. If you think the code isn't clear by itself (I think it's fine as is) you could make it more obvious by doing something like:

```
const sourceCodeIntegrationEnabled = ...
const apiKeyProvided = ...
const isGitRepo = ...

if (sourceCodeIntegrationEnabled && apiKeyProvided && isGitRepo) {
...
```

nhinsch reviewed (approved):
LGTM, just left one nit!
